crates/hyperdrive-math/src/short/max.rs

```diff
@@ -115,33 +116,73 @@ impl State {
             open_vault_share_price,
             maybe_conservative_price,
         );
+        let mut best_valid_max_bond_amount = max_bond_amount;
+
+        // To avoid the case where Newton's method overshoots and stays on
+        // the invalid side of the optimization equation (i.e., when deposit > budget),
+        // we artificially set the target budget to be less than the actual budget.
+        // We use the minimum transaction amount here to avoid the underflow
+        // where budget is less than this bias.
+        let target_budget = budget - self.minimum_transaction_amount();
         for _ in 0..maybe_max_iterations.unwrap_or(7) {
             let deposit = match self.calculate_open_short(
                 max_bond_amount,
                 spot_price,
                 open_vault_share_price,
             ) {
                 Ok(d) => d,
-                Err(_) => return max_bond_amount,
+                Err(_) => {
+                    // The pool is insolvent for the guess at this point.
+                    // We use the absolute max bond amount and deposit
+                    // for the next guess iteration
+                    max_bond_amount = absolute_max_bond_amount;
+                    absolute_max_deposit
+                }
             };
-            max_bond_amount += (budget - deposit)
-                / self.short_deposit_derivative(
-                    max_bond_amount,
-                    spot_price,
-                    open_vault_share_price,
-                );
+
+            let derivative =
+                self.short_deposit_derivative(max_bond_amount, spot_price, open_vault_share_price);
```

Comment:
Can we move this closer to where it is used? If we move this below `// Iteratively update max_bond_amount via newton's method`, it would be easier to read IMO.